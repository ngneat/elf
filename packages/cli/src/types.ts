import { SourceFile } from 'ts-morph';

export interface Options {
  storeName: string;
  features: Array<Features>;
  crud: Array<
    'addEntities' | 'updateEntities' | 'deleteEntities' | 'setEntities'
  >;
  path: string;
  idKey: string;
  template?: 'class' | 'functions';
  inlineStoreInClass?: boolean;
  hooks?: Array<Hooks>;
}

export interface Hooks {
  post?(options: {
    sourceFile: SourceFile;
    options: Omit<Options, 'hooks'>;
    repoName: string;
  }): void;
}

export interface GlobalConfig {
  cli?: {
    repoTemplate?: Options['template'];
    inlineStoreInClass?: Options['inlineStoreInClass'];
    idKey?: Options['idKey'];
    repoLibrary?: string;
    plugins?: string[];
    fuzzypath?: {
      rootPath?: string;
      excludeFilter?: Function;
      excludePath?: Function;
    };
  };
}

export type Features = typeof baseFeatures[number]['value'];

export const baseClassStorePlaces = [
  { name: 'Outside of a class', value: false },
  { name: 'Inside a class constructor', value: true },
] as const;

export const baseFeatures = [
  { name: 'Props', value: 'withProps' },
  { name: 'Entities', value: 'withEntities' },
  {
    name: 'UIEntities',
    value: 'withUIEntities',
  },
  {
    name: 'Active Id',
    value: 'withActiveId',
  },
  {
    name: 'Active Ids',
    value: 'withActiveIds',
  },
  {
    name: 'Requests Cache',
    value: 'withRequestsCache',
  },
  {
    name: 'Requests Status',
    value: 'withRequestsStatus',
  },
] as const;

export const DEFAULT_ID_KEY = 'id';
