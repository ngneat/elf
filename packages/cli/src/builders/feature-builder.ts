import {
  ClassDeclaration,
  SourceFile,
  StructureKind,
  ConstructorDeclaration,
} from 'ts-morph';
import { Features, Options } from '../types';
import { coerceArray, names, resolveStoreVariableName } from '../utils';

// @ts-ignore
import * as pluralize from 'pluralize';
import { CallExpression } from 'typescript';

export abstract class FeatureBuilder {
  static supports(featureName: Features): boolean {
    return false;
  }

  storeName = this.options.storeName;
  idKey = this.options.idKey;
  storeSingularNames = names(pluralize.singular(this.storeName));
  storeNames = names(this.storeName);
  isFunctionsTpl =
    !this.options.template || this.options.template === 'functions';
  isStoreInlinedInClass =
    !this.isFunctionsTpl && this.options.inlineStoreInClass;
  storeVariableName = resolveStoreVariableName(
    this.options.template,
    this.storeNames,
    this.isStoreInlinedInClass
  );
  repoConstructor: ConstructorDeclaration | undefined;

  constructor(
    protected sourceFile: SourceFile,
    protected repo: ClassDeclaration,
    protected options: Options
  ) {
    if (this.isStoreInlinedInClass) {
      this.repoConstructor = this.repo.getConstructors()[0];
    }
  }

  abstract run(): void;

  abstract getPropsFactory(): CallExpression;

  addImport(name: string | string[], moduleSpecifier = '@ngneat/elf') {
    const importDecl = this.sourceFile.getImportDeclaration(moduleSpecifier);

    if (!importDecl) {
      this.sourceFile.insertImportDeclaration(this.getLastImportIndex() + 1, {
        moduleSpecifier,
        namedImports: coerceArray(name).map((name) => ({
          kind: StructureKind.ImportSpecifier,
          name,
        })),
      });
    } else {
      coerceArray(name).forEach((v) => importDecl.addNamedImport(v));
    }
  }

  getLastImportIndex() {
    const imports = this.sourceFile.getImportDeclarations();

    return imports[imports.length - 1].getChildIndex();
  }
}
