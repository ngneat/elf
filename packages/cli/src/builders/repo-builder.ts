import {
  ClassDeclaration,
  ConstructorDeclaration,
  printNode,
  Project,
  QuoteKind,
  Scope,
  SourceFile,
  StructureKind,
  SyntaxKind,
  VariableDeclarationKind,
} from 'ts-morph';
import { CallExpression, factory, ScriptTarget } from 'typescript';
import { Options } from '../types';
import { names, resolveStoreVariableName } from '../utils';
import { ActiveIdBuilder } from './active-id.builder';
import { ActiveIdsBuilder } from './active-ids.builder';
import { EntitiesBuilder } from './entities.builder';
import { PropsBuilder } from './props.builder';
import { RequestsCacheBuilder } from './requests-cache.builder';
import { RequestsStatusBuilder } from './requests-status.builder';
import { UIEntitiesBuilder } from './ui-entities.builder';

export function createRepo(options: Options) {
  const { storeName } = options;
  const storeNames = names(storeName);
  const isFunctionTpl = !options.template || options.template === 'functions';
  const isStoreInlinedInClass = !isFunctionTpl && options.inlineStoreInClass;

  const project = new Project({
    manipulationSettings: {
      quoteKind: QuoteKind.Single,
    },
    compilerOptions: {
      target: ScriptTarget.ES2015,
    },
  });

  const sourceFile = project.createSourceFile(`repo.ts`, ``);

  const repoName = `${storeNames.className}Repository`;
  const repoClassDec = sourceFile.addClass({
    name: repoName,
    isExported: true,
  });
  let repoClassDecConstructor: ConstructorDeclaration | undefined;

  if (isStoreInlinedInClass) {
    repoClassDec.addConstructor();
    repoClassDecConstructor = repoClassDec.getConstructors()[0];
  }

  sourceFile.addImportDeclaration({
    moduleSpecifier: '@ngneat/elf',
    namedImports: ['createStore'].map((name) => ({
      kind: StructureKind.ImportSpecifier,
      name,
    })),
  });

  const builders = [
    RequestsCacheBuilder,
    RequestsStatusBuilder,
    ActiveIdBuilder,
    ActiveIdsBuilder,
    EntitiesBuilder,
    PropsBuilder,
    UIEntitiesBuilder,
  ];

  const propsFactories: CallExpression[] = [];

  for (const feature of options.features) {
    for (const Builder of builders) {
      if (Builder.supports(feature)) {
        const instance = new Builder(sourceFile, repoClassDec, options);
        instance.run();
        propsFactories.push(instance.getPropsFactory());
      }
    }
  }

  const storeOpts = factory.createIdentifier(
    `{ name: '${storeNames.propertyName}' }`
  );

  const store = factory.createCallExpression(
    factory.createIdentifier('createStore'),
    undefined,
    [storeOpts, ...propsFactories]
  );

  if (isStoreInlinedInClass && repoClassDecConstructor) {
    addInlineStoreToRepoClass({
      repoClassDec,
      repoClassDecConstructor,
      options,
      store,
      storeNames,
    });
  } else {
    addStoreToRepo({
      repoClassDec,
      options,
      store,
      storeNames,
      isFunctionTpl,
      sourceFile,
    });

    if (isFunctionTpl) {
      toFunctions(sourceFile, repoClassDec);
    }
  }

  if (options.hooks) {
    options.hooks.forEach((h) => h.post?.({ sourceFile, repoName, options }));
  }

  sourceFile.formatText({ indentSize: 2 });

  return sourceFile.getText();
}

function addStoreToRepo({
  repoClassDec: classDec,
  sourceFile,
  options,
  storeNames,
  store,
  isFunctionTpl,
}: {
  repoClassDec: ClassDeclaration;
  sourceFile: SourceFile;
  options: Options;
  storeNames: ReturnType<typeof names>;
  store: CallExpression;
  isFunctionTpl: boolean;
}) {
  const repoPosition = classDec.getChildIndex();

  sourceFile.insertVariableStatement(repoPosition, {
    declarationKind: VariableDeclarationKind.Const,
    isExported: isFunctionTpl,
    declarations: [
      {
        name: resolveStoreVariableName(options.template, storeNames),
        initializer: printNode(store),
      },
    ],
  });
}

function addInlineStoreToRepoClass({
  repoClassDec: classDec,
  repoClassDecConstructor: constructorDec,
  options,
  storeNames,
  store,
}: {
  repoClassDec: ClassDeclaration;
  repoClassDecConstructor: ConstructorDeclaration;
  options: Options;
  storeNames: ReturnType<typeof names>;
  store: CallExpression;
}) {
  const storeName = resolveStoreVariableName(
    options.template,
    storeNames,
    true
  );
  const { propertyIndex, methodIndex } = getPositionsOfInlineStoreDeclarations(
    classDec,
    constructorDec
  );

  const createStoreMethodName = 'createStore';

  classDec.insertMethod(methodIndex, {
    name: createStoreMethodName,
    returnType: `typeof store`,
    scope: Scope.Private,
    statements: (writer) => {
      writer.writeLine(`const store = ${printNode(store)};`);
      writer.blankLine();
      writer.writeLine(`return store;`);
    },
  });

  constructorDec.insertStatements(
    0,
    `${storeName} = this.${createStoreMethodName}();`
  );

  const storeProperty = classDec.insertProperty(propertyIndex, {
    name: `${resolveStoreVariableName(options.template, storeNames)}`,
    scope: Scope.Private,
  });

  if (propertyIndex > 0) {
    storeProperty?.prependWhitespace('\n');
  }
}

function toFunctions(sourceFile: SourceFile, classDec: ClassDeclaration) {
  const exported: string[] = [];

  classDec?.getProperties().forEach((p) => {
    exported.push(`export const ${p.getText()}`);
  });

  classDec?.getMethods().forEach((m) => {
    exported.push(`export function ${m.getText()}`);
  });

  classDec?.remove();

  sourceFile.replaceWithText(
    `${sourceFile.getText()}\n ${exported.join('\n\n')}`
  );
}

function getPositionsOfInlineStoreDeclarations(
  classDec: ClassDeclaration,
  constructorDec: ConstructorDeclaration
) {
  const lastPropertyIndex = classDec
    .getLastChildByKind(SyntaxKind.PropertyDeclaration)
    ?.getChildIndex();
  const lastMethodIndex = classDec
    .getLastChildByKind(SyntaxKind.MethodDeclaration)
    ?.getChildIndex();

  return {
    methodIndex: (lastMethodIndex ?? constructorDec.getChildIndex()) + 1,
    propertyIndex: lastPropertyIndex
      ? lastPropertyIndex + 1
      : constructorDec.getChildIndex(),
  };
}
