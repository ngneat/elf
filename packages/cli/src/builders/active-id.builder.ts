import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { StructureKind, ClassMemberStructures } from 'ts-morph';
import { factory } from 'typescript';

export class ActiveIdBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withActiveId';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withActiveId'),
      undefined,
      []
    );
  }

  run() {
    this.addImport(
      ['withActiveId', 'selectActiveEntity', 'setActiveId'],
      '@ngneat/elf-entities'
    );

    const initializer = `${this.storeVariableName}.pipe(selectActiveEntity())`;
    const memberData: ClassMemberStructures = {
      name: `active${this.storeSingularNames.className}$`,
      kind: StructureKind.Property,
    };

    if (this.isStoreInlinedInClass) {
      this.repo.insertProperty(0, {
        ...memberData,
        type: `Observable<${this.storeSingularNames.className} | undefined>`,
      });

      this.repoConstructor?.addStatements(
        `this.${memberData.name} = ${initializer};`
      );
    } else {
      this.repo.insertMember(0, {
        ...memberData,
        initializer,
      });
    }

    this.repo.addMember({
      kind: StructureKind.Method,
      name: this.isFunctionsTpl
        ? `setActive${this.storeNames.className}Id`
        : `setActiveId`,
      parameters: [
        {
          name: 'id',
          type: `${this.storeSingularNames.className}['${this.idKey}']`,
        },
      ],
      statements: [`${this.storeVariableName}.update(setActiveId(id));`],
    });
  }
}
