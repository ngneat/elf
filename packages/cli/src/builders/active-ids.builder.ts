import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { StructureKind, ClassMemberStructures } from 'ts-morph';
import { factory } from 'typescript';

export class ActiveIdsBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withActiveIds';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withActiveIds'),
      undefined,
      []
    );
  }

  run() {
    this.addImport(
      ['withActiveIds', 'selectActiveEntities', 'toggleActiveIds'],
      '@ngneat/elf-entities'
    );

    const initializer = `${this.storeVariableName}.pipe(selectActiveEntities())`;
    const memberData: ClassMemberStructures = {
      name: `active${this.storeNames.className}$`,
      kind: StructureKind.Property,
    };

    if (this.isStoreInlinedInClass) {
      this.repo.insertProperty(0, {
        ...memberData,
        type: `Observable<${this.storeSingularNames.className}[]>`,
      });

      this.repoConstructor?.addStatements(
        `this.${memberData.name} = ${initializer};`
      );
    } else {
      this.repo.insertMember(0, {
        ...memberData,
        initializer,
      });
    }

    this.repo.addMember({
      kind: StructureKind.Method,
      name: this.isFunctionsTpl
        ? `toggleActive${this.storeNames.className}Ids`
        : `toggleActiveIds`,
      parameters: [
        {
          name: 'ids',
          type: `Array<${this.storeSingularNames.className}['${this.idKey}']>`,
        },
      ],
      statements: [`${this.storeVariableName}.update(toggleActiveIds(ids));`],
    });
  }
}
