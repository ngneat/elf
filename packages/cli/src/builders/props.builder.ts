import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { factory } from 'typescript';

export class PropsBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withProps';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withProps'),
      [
        factory.createTypeReferenceNode(
          factory.createIdentifier(`${this.storeNames.className}Props`),
          undefined
        ),
      ],
      [factory.createObjectLiteralExpression([], false)]
    );
  }

  run() {
    this.addImport(['withProps']);

    const decl = this.sourceFile.insertInterface(
      this.getLastImportIndex() + 1,
      {
        name: `${this.storeNames.className}Props`,
        isExported: true,
      }
    );

    decl.replaceWithText(
      `// eslint-disable-next-line @typescript-eslint/no-empty-interface\n${decl.getText()}`
    );
  }
}
