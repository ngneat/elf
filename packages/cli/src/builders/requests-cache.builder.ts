import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { factory } from 'typescript';
import { StructureKind } from 'ts-morph';

export class RequestsCacheBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withRequestsCache';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier(`withRequestsCache<'${this.storeName}'>`),
      undefined,
      []
    );
  }

  run() {
    this.addImport(['withRequestsCache'], '@ngneat/elf-requests');
  }
}
