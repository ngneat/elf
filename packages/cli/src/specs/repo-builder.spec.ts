import { createRepo } from '../builders/repo-builder';
import { baseFeatures } from '../types';

describe('Repo Builder', () => {
  it('withProps', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withProps'],
      crud: [],
      template: 'class',
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withProps');
  });

  it('withProps in a class', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withProps'],
      crud: [],
      template: 'class',
      inlineStoreInClass: true,
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withProps in a class');
  });

  it('withRequestsCache', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withRequestsCache'],
      crud: [],
      storeName: 'todos',
      path: '',
      template: 'class',
    });

    expect(output).toMatchSnapshot('withRequestsCache');
  });

  it('withRequestsCache in a class', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withRequestsCache'],
      crud: [],
      template: 'class',
      inlineStoreInClass: true,
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withRequestsCache in a class');
  });

  it('withRequestsStatus', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withRequestsStatus'],
      crud: [],
      storeName: 'todos',
      path: '',
      template: 'class',
    });

    expect(output).toMatchSnapshot('withRequestsStatus');
  });

  it('withRequestsStatus in a class', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withRequestsStatus'],
      crud: [],
      template: 'class',
      inlineStoreInClass: true,
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withRequestsStatus in a class');
  });

  it('withActiveId', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withActiveId'],
      crud: [],
      storeName: 'todos',
      path: '',
      template: 'class',
    });

    expect(output).toMatchSnapshot('withActiveId');
  });

  it('withActiveId in a class', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withActiveId'],
      crud: [],
      template: 'class',
      inlineStoreInClass: true,
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withActiveId in a class');
  });

  it('withActiveIds', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withActiveIds'],
      crud: [],
      storeName: 'todos',
      path: '',
      template: 'class',
    });

    expect(output).toMatchSnapshot('withActiveIds');
  });

  it('withActiveIds in a class', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withActiveIds'],
      crud: [],
      template: 'class',
      inlineStoreInClass: true,
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withActiveIds in a class');
  });

  it('withEntities', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withEntities'],
      crud: ['addEntities', 'updateEntities', 'deleteEntities'],
      storeName: 'todos',
      path: '',
      template: 'class',
    });

    expect(output).toMatchSnapshot('withEntities');
  });

  it('withEntities in a class', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withEntities'],
      crud: [],
      template: 'class',
      inlineStoreInClass: true,
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withEntities in a class');
  });

  it('withEntities idKey', () => {
    const output = createRepo({
      idKey: '_id',
      features: ['withEntities'],
      crud: ['addEntities', 'updateEntities', 'deleteEntities'],
      storeName: 'todos',
      path: '',
      template: 'class',
    });

    expect(output).toMatchSnapshot('withEntities');
  });

  it('withEntities idKey in a class', () => {
    const output = createRepo({
      idKey: '_id',
      features: ['withEntities'],
      crud: ['addEntities', 'updateEntities', 'deleteEntities'],
      template: 'class',
      inlineStoreInClass: true,
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withEntities in a class');
  });

  it('withUIEntities', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withUIEntities'],
      crud: [],
      storeName: 'todos',
      path: '',
      template: 'class',
    });

    expect(output).toMatchSnapshot('withUIEntities');
  });

  it('withUIEntities in a class', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withUIEntities'],
      crud: [],
      template: 'class',
      inlineStoreInClass: true,
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withUIEntities in a class');
  });

  it('all', () => {
    const output = createRepo({
      idKey: 'id',
      features: baseFeatures.map(({ value }) => value),
      crud: ['deleteEntities', 'addEntities', 'updateEntities'],
      storeName: 'todos',
      path: '',
      template: 'class',
    });

    expect(output).toMatchSnapshot('all');
  });

  it('all in a class', () => {
    const output = createRepo({
      idKey: 'id',
      features: baseFeatures.map(({ value }) => value),
      crud: ['deleteEntities', 'addEntities', 'updateEntities'],
      template: 'class',
      inlineStoreInClass: true,
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('all in a class');
  });

  it('should support function template', () => {
    const output = createRepo({
      idKey: 'id',
      features: baseFeatures.map(({ value }) => value),
      crud: ['deleteEntities', 'addEntities', 'updateEntities'],
      storeName: 'todos',
      path: '',
      template: 'functions',
    });

    expect(output).toMatchSnapshot('all');
  });
});
