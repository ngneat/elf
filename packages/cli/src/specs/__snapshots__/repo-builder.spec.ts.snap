// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repo Builder all in a class: all in a class 1`] = `
"import { createStore, withProps } from '@ngneat/elf';
import { withEntities, selectAllEntities, deleteEntities, addEntities, updateEntities, withUIEntities, withActiveId, selectActiveEntity, setActiveId, withActiveIds, selectActiveEntities, toggleActiveIds } from '@ngneat/elf-entities';
import { Observable } from 'rxjs';
import { withRequestsCache, withRequestsStatus } from '@ngneat/elf-requests';

export interface TodoUI {
  id: number;
}

export interface Todo {
  id: number;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TodosProps {
}

export class TodosRepository {
  activeTodos$: Observable<Todo[]>;
  activeTodo$: Observable<Todo | undefined>;
  todos$: Observable<Todo[]>;

  private store;

  constructor() {
    this.store = this.createStore();
    this.todos$ = this.store.pipe(selectAllEntities());
    this.activeTodo$ = this.store.pipe(selectActiveEntity());
    this.activeTodos$ = this.store.pipe(selectActiveEntities());
  }

  deleteTodo(id: Todo['id']) {
    this.store.update(deleteEntities(id));
  }

  addTodo(todo: Todo) {
    this.store.update(addEntities(todo));
  }

  updateTodo(id: Todo['id'], todo: Partial<Todo>) {
    this.store.update(updateEntities(id, todo));
  }

  setActiveId(id: Todo['id']) {
    this.store.update(setActiveId(id));
  }

  toggleActiveIds(ids: Array<Todo['id']>) {
    this.store.update(toggleActiveIds(ids));
  }

  private createStore(): typeof store {
    const store = createStore({ name: 'todos' }, withProps<TodosProps>({}), withEntities<Todo>(), withUIEntities<TodoUI>(), withActiveId(), withActiveIds(), withRequestsCache<'todos'>(), withRequestsStatus<'todos'>());

    return store;
  }
}
"
`;

exports[`Repo Builder all: all 1`] = `
"import { createStore, withProps } from '@ngneat/elf';
import { withEntities, selectAllEntities, deleteEntities, addEntities, updateEntities, withUIEntities, withActiveId, selectActiveEntity, setActiveId, withActiveIds, selectActiveEntities, toggleActiveIds } from '@ngneat/elf-entities';
import { withRequestsCache, withRequestsStatus } from '@ngneat/elf-requests';

export interface TodoUI {
  id: number;
}

export interface Todo {
  id: number;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TodosProps {
}

const store = createStore({ name: 'todos' }, withProps<TodosProps>({}), withEntities<Todo>(), withUIEntities<TodoUI>(), withActiveId(), withActiveIds(), withRequestsCache<'todos'>(), withRequestsStatus<'todos'>());

export class TodosRepository {
  activeTodos$ = store.pipe(selectActiveEntities());
  activeTodo$ = store.pipe(selectActiveEntity());
  todos$ = store.pipe(selectAllEntities());

  deleteTodo(id: Todo['id']) {
    store.update(deleteEntities(id));
  }

  addTodo(todo: Todo) {
    store.update(addEntities(todo));
  }

  updateTodo(id: Todo['id'], todo: Partial<Todo>) {
    store.update(updateEntities(id, todo));
  }

  setActiveId(id: Todo['id']) {
    store.update(setActiveId(id));
  }

  toggleActiveIds(ids: Array<Todo['id']>) {
    store.update(toggleActiveIds(ids));
  }
}
"
`;

exports[`Repo Builder should support function template: all 1`] = `
"import { createStore, withProps } from '@ngneat/elf';
import { withEntities, selectAllEntities, deleteEntities, addEntities, updateEntities, withUIEntities, withActiveId, selectActiveEntity, setActiveId, withActiveIds, selectActiveEntities, toggleActiveIds } from '@ngneat/elf-entities';
import { withRequestsCache, withRequestsStatus } from '@ngneat/elf-requests';

export interface TodoUI {
  id: number;
}

export interface Todo {
  id: number;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TodosProps {
}

export const todosStore = createStore({ name: 'todos' }, withProps<TodosProps>({}), withEntities<Todo>(), withUIEntities<TodoUI>(), withActiveId(), withActiveIds(), withRequestsCache<'todos'>(), withRequestsStatus<'todos'>());

export const activeTodos$ = todosStore.pipe(selectActiveEntities());

export const activeTodo$ = todosStore.pipe(selectActiveEntity());

export const todos$ = todosStore.pipe(selectAllEntities());

export function deleteTodo(id: Todo['id']) {
  todosStore.update(deleteEntities(id));
}

export function addTodo(todo: Todo) {
  todosStore.update(addEntities(todo));
}

export function updateTodo(id: Todo['id'], todo: Partial<Todo>) {
  todosStore.update(updateEntities(id, todo));
}

export function setActiveTodosId(id: Todo['id']) {
  todosStore.update(setActiveId(id));
}

export function toggleActiveTodosIds(ids: Array<Todo['id']>) {
  todosStore.update(toggleActiveIds(ids));
}
"
`;

exports[`Repo Builder withActiveId in a class: withActiveId in a class 1`] = `
"import { createStore } from '@ngneat/elf';
import { withActiveId, selectActiveEntity, setActiveId } from '@ngneat/elf-entities';

export class TodosRepository {
  activeTodo$: Observable<Todo | undefined>;

  private store;

  constructor() {
    this.store = this.createStore();
    this.activeTodo$ = this.store.pipe(selectActiveEntity());
  }

  setActiveId(id: Todo['id']) {
    this.store.update(setActiveId(id));
  }

  private createStore(): typeof store {
    const store = createStore({ name: 'todos' }, withActiveId());

    return store;
  }
}
"
`;

exports[`Repo Builder withActiveId: withActiveId 1`] = `
"import { createStore } from '@ngneat/elf';
import { withActiveId, selectActiveEntity, setActiveId } from '@ngneat/elf-entities';

const store = createStore({ name: 'todos' }, withActiveId());

export class TodosRepository {
  activeTodo$ = store.pipe(selectActiveEntity());

  setActiveId(id: Todo['id']) {
    store.update(setActiveId(id));
  }
}
"
`;

exports[`Repo Builder withActiveIds in a class: withActiveIds in a class 1`] = `
"import { createStore } from '@ngneat/elf';
import { withActiveIds, selectActiveEntities, toggleActiveIds } from '@ngneat/elf-entities';

export class TodosRepository {
  activeTodos$: Observable<Todo[]>;

  private store;

  constructor() {
    this.store = this.createStore();
    this.activeTodos$ = this.store.pipe(selectActiveEntities());
  }

  toggleActiveIds(ids: Array<Todo['id']>) {
    this.store.update(toggleActiveIds(ids));
  }

  private createStore(): typeof store {
    const store = createStore({ name: 'todos' }, withActiveIds());

    return store;
  }
}
"
`;

exports[`Repo Builder withActiveIds: withActiveIds 1`] = `
"import { createStore } from '@ngneat/elf';
import { withActiveIds, selectActiveEntities, toggleActiveIds } from '@ngneat/elf-entities';

const store = createStore({ name: 'todos' }, withActiveIds());

export class TodosRepository {
  activeTodos$ = store.pipe(selectActiveEntities());

  toggleActiveIds(ids: Array<Todo['id']>) {
    store.update(toggleActiveIds(ids));
  }
}
"
`;

exports[`Repo Builder withEntities idKey in a class: withEntities in a class 1`] = `
"import { createStore } from '@ngneat/elf';
import { withEntities, selectAllEntities, addEntities, updateEntities, deleteEntities } from '@ngneat/elf-entities';
import { Observable } from 'rxjs';

export interface Todo {
  _id: number;
}

export class TodosRepository {
  todos$: Observable<Todo[]>;

  private store;

  constructor() {
    this.store = this.createStore();
    this.todos$ = this.store.pipe(selectAllEntities());
  }

  addTodo(todo: Todo) {
    this.store.update(addEntities(todo));
  }

  updateTodo(id: Todo['_id'], todo: Partial<Todo>) {
    this.store.update(updateEntities(id, todo));
  }

  deleteTodo(id: Todo['_id']) {
    this.store.update(deleteEntities(id));
  }

  private createStore(): typeof store {
    const store = createStore({ name: 'todos' }, withEntities<Todo, '_id'>({ idKey: '_id' }));

    return store;
  }
}
"
`;

exports[`Repo Builder withEntities idKey: withEntities 1`] = `
"import { createStore } from '@ngneat/elf';
import { withEntities, selectAllEntities, addEntities, updateEntities, deleteEntities } from '@ngneat/elf-entities';

export interface Todo {
  _id: number;
}

const store = createStore({ name: 'todos' }, withEntities<Todo, '_id'>({ idKey: '_id' }));

export class TodosRepository {
  todos$ = store.pipe(selectAllEntities());

  addTodo(todo: Todo) {
    store.update(addEntities(todo));
  }

  updateTodo(id: Todo['_id'], todo: Partial<Todo>) {
    store.update(updateEntities(id, todo));
  }

  deleteTodo(id: Todo['_id']) {
    store.update(deleteEntities(id));
  }
}
"
`;

exports[`Repo Builder withEntities in a class: withEntities in a class 1`] = `
"import { createStore } from '@ngneat/elf';
import { withEntities, selectAllEntities } from '@ngneat/elf-entities';
import { Observable } from 'rxjs';

export interface Todo {
  id: number;
}

export class TodosRepository {
  todos$: Observable<Todo[]>;

  private store;

  constructor() {
    this.store = this.createStore();
    this.todos$ = this.store.pipe(selectAllEntities());
  }

  private createStore(): typeof store {
    const store = createStore({ name: 'todos' }, withEntities<Todo>());

    return store;
  }
}
"
`;

exports[`Repo Builder withEntities: withEntities 1`] = `
"import { createStore } from '@ngneat/elf';
import { withEntities, selectAllEntities, addEntities, updateEntities, deleteEntities } from '@ngneat/elf-entities';

export interface Todo {
  id: number;
}

const store = createStore({ name: 'todos' }, withEntities<Todo>());

export class TodosRepository {
  todos$ = store.pipe(selectAllEntities());

  addTodo(todo: Todo) {
    store.update(addEntities(todo));
  }

  updateTodo(id: Todo['id'], todo: Partial<Todo>) {
    store.update(updateEntities(id, todo));
  }

  deleteTodo(id: Todo['id']) {
    store.update(deleteEntities(id));
  }
}
"
`;

exports[`Repo Builder withProps in a class: withProps in a class 1`] = `
"import { createStore, withProps } from '@ngneat/elf';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TodosProps {
}

export class TodosRepository {
  private store;

  constructor() {
    this.store = this.createStore();
  }

  private createStore(): typeof store {
    const store = createStore({ name: 'todos' }, withProps<TodosProps>({}));

    return store;
  }
}
"
`;

exports[`Repo Builder withProps: withProps 1`] = `
"import { createStore, withProps } from '@ngneat/elf';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TodosProps {
}

const store = createStore({ name: 'todos' }, withProps<TodosProps>({}));

export class TodosRepository {
}
"
`;

exports[`Repo Builder withRequestsCache in a class: withRequestsCache in a class 1`] = `
"import { createStore } from '@ngneat/elf';
import { withRequestsCache } from '@ngneat/elf-requests';

export class TodosRepository {
  private store;

  constructor() {
    this.store = this.createStore();
  }

  private createStore(): typeof store {
    const store = createStore({ name: 'todos' }, withRequestsCache<'todos'>());

    return store;
  }
}
"
`;

exports[`Repo Builder withRequestsCache: withRequestsCache 1`] = `
"import { createStore } from '@ngneat/elf';
import { withRequestsCache } from '@ngneat/elf-requests';

const store = createStore({ name: 'todos' }, withRequestsCache<'todos'>());

export class TodosRepository {
}
"
`;

exports[`Repo Builder withRequestsStatus in a class: withRequestsStatus in a class 1`] = `
"import { createStore } from '@ngneat/elf';
import { withRequestsStatus } from '@ngneat/elf-requests';

export class TodosRepository {
  private store;

  constructor() {
    this.store = this.createStore();
  }

  private createStore(): typeof store {
    const store = createStore({ name: 'todos' }, withRequestsStatus<'todos'>());

    return store;
  }
}
"
`;

exports[`Repo Builder withRequestsStatus: withRequestsStatus 1`] = `
"import { createStore } from '@ngneat/elf';
import { withRequestsStatus } from '@ngneat/elf-requests';

const store = createStore({ name: 'todos' }, withRequestsStatus<'todos'>());

export class TodosRepository {
}
"
`;

exports[`Repo Builder withUIEntities in a class: withUIEntities in a class 1`] = `
"import { createStore } from '@ngneat/elf';
import { withUIEntities } from '@ngneat/elf-entities';

export interface TodoUI {
  id: number;
}

export class TodosRepository {
  private store;

  constructor() {
    this.store = this.createStore();
  }

  private createStore(): typeof store {
    const store = createStore({ name: 'todos' }, withUIEntities<TodoUI>());

    return store;
  }
}
"
`;

exports[`Repo Builder withUIEntities: withUIEntities 1`] = `
"import { createStore } from '@ngneat/elf';
import { withUIEntities } from '@ngneat/elf-entities';

export interface TodoUI {
  id: number;
}

const store = createStore({ name: 'todos' }, withUIEntities<TodoUI>());

export class TodosRepository {
}
"
`;
